### ISSUE TITLE
More illustration figures on different types of path identifiers?

### ISSUE THREAD

**Issue by yfmascgy** [quicwg/multipath#1]
There is one part I am afraid people might feel a little bit hard to follow: the three types of path identifiers in the PATH_STATUS frame. Those three types of path identifiers were introduced in order to be compatible with zero-length CID use cases, so I think we might need two additional example figures (like figure 3 in the current draft), one for (non-zero length CID, zero-length CID) and the other one for (zero-length CID & zero-length CID). Do you think that is a good idea? I can work on those figures if it doesn't complicate the flow of the draft.

**Comment by yfmascgy** [comment]
@huitema I am editing the path closing figures with zero-length CID and single PN. Is it similar to what you have in mind?

~~~
  Client                                                      Server

  (client tells server to abandon a path)
  1-RTT[X]: DCID=S2 PATH_ABANDON[path_id_type=1, path_id=2]->
                             (server tells client to abandon a path)
                <-1-RTT[Y]: PATH_ABANDON[path_id_type=0, path_id=2],
                                                 ACK_MP[Seq=0, PN=X]
  (client abandons the path that it is using)
  1-RTT[X+1]: DCID=S3 RETIRE_CONNECTION_ID[2],
                            ACK_MP[Seq=0, PN=Y]->
                                 <-1-RTT[Y+1]: ACK_MP[Seq=0, PN=X+1]
  (client waits for idle timeout to release resources)

~~~
{: #fig-example-path-close2 title="Example of closing a path when the server chooses to use zero-length CID"}

~~~
  Client                                                     Server

  (client tells server to abandon a path)
  1-RTT[X]: DCID=S2 PATH_ABANDON[path_id_type=2]->
                            (server tells client to abandon a path)
                          <-1-RTT[Y]: PATH_ABANDON[path_id_type=2],
                                                ACK_MP[Seq=0, PN=X]
  1-RTT[X+1]: ACK_MP[Seq=0, PN=Y]->
  (client waits for idle timeout to release resources)
               (server waits for idle timeout to release resources)

~~~
{: #fig-example-path-close3 title="Example of closing a path when the both client and server choose to use zero-length CIDs."}

**Comment by huitema** [comment]
In the case of single number space, we use ACK, not ACK_MP.  But in fact you do not need to draw the ACKs there, there is no special processing of ACKs required. Also, there is no need to wait for an idle timeout, which typically is 10 to 30 seconds. That's too large. 

Do we really need to send "abandon path" in both directions? If I think about it, the real concern is that client will send a PATH_ABANDON message, but some time later, due to differences in delay, etc., the server receive data on the same four tuples.
How does the server know whether it should create a new path or not?

In the single space case, the logical test may be to check the sequence numbers of the packet. If the sequence number of the new packet is greater than that of the path_abandon, then the client is reactivating a path that was abandoned previously. Maybe the client connected again to the same network, after some disconnected interval. The server can start sending on the reactivated path. If the packet number is lower than that of the path abandon, then the server should not reactivate the path.

In the multiple space case, if the CID is defined, there is no ambiguity.

**Comment by yfmascgy** [comment]
> In the case of single number space, we use ACK, not ACK_MP. But in fact you do not need to draw the ACKs there, there is no special processing of ACKs required. Also, there is no need to wait for an idle timeout, which typically is 10 to 30 seconds. That's too large.

ACK works for me. I think we might also want to add a sentence in the Sec. of "one packet number space" explicitly specifying that "for single PN, use ACK instead of ACK_MP", and change some texts in the ACK_MP frame section that is related to zero-length CID (the current draft says "if the endpoint receives 1-RTT packets with 0-length Connection ID, it SHOULD use Packet Number Space Identifier 0 in ACK_MP frames". ).

> 
> Do we really need to send "abandon path" in both directions? If I think about it, the real concern is that client will send a PATH_ABANDON message, but some time later, due to differences in delay, etc., the server receive data on the same four tuples. How does the server know whether it should create a new path or not?

I think in single PN, sending "abandon path" in both directions is helpful. When the client sends "PATH_ABANDON", it promises not to send non-probing packets in the current activity cycle. It could send probing packets to reactivate the path a moment later if the client thinks the network environment has changed. After sending out "PATH_ABANDON", the client should still be able to receive packets from the server for a while on that path because there may be the server's inflight packets on that path. It was not until the client sees the server's "PATH_ABANDON", the client then knows that no more server's data is to be expected except for some out-of-order packets. So after receiving "PATH_ABANDON", the endpoint can start preparing to free memories and clear path stats allocated for that path. The time interval from seeing "PATH_ABANDON" to actually releasing relevant resources is going to be an important parameter, and yes, 10-30 seconds is too long. What would be a more appropriate number for this interval? 

> 
> In the single space case, the logical test may be to check the sequence numbers of the packet. If the sequence number of the new packet is greater than that of the path_abandon, then the client is reactivating a path that was abandoned previously. Maybe the client connected again to the same network, after some disconnected interval. The server can start sending on the reactivated path. If the packet number is lower than that of the path abandon, then the server should not reactivate the path.

Does it imply that we need to track the packet number of the PATH_ABANDON for a path that is closed? When a path is in a closed state, the server may reject any newly received packets unless it is a probing packet (back to our path initialization process). In this way, it should be OK to forget any info related to that path in the last activity cycle.

> 
> In the multiple space case, if the CID is defined, there is no ambiguity.

**Comment by mirjak** [comment]
addressed by PR #104
